Consolidate element instruction code

crisbeto opened the pull request:
Includes the following changes that consolidate some shared logic:

### refactor(core): consolidate first create pass 
The first create pass for elements and containers was identical. These changes consolidate it to reduce code duplication.

### refactor(core): consolidate element start logic 
The logic for the `elementStart` and `elementContainerStart` instructions was identical. These changes consolidate it into a single base instruction.

### refactor(core): consolidate element end instruction logic 
There was some identical logic between the `elementEnd` and `elementContainerEnd` instructions. These changes consolidate it.

### refactor(core): consolidate logic to determine whether node can be hydrated
Several instructions were repeating the logic that checks if a specific can be hydrated. These changes move it into a common location.

crisbeto committed:
refactor(core): consolidate first create pass

The first create pass for elements and containers was identical. These changes consolidate it to reduce code duplication.

crisbeto committed:
refactor(core): consolidate element start logic

The logic for the `elementStart` and `elementContainerStart` instructions was identical. These changes consolidate it into a single base instruction.

pkozlowski-opensource commented on packages/core/src/render3/view/elements.ts:
```diff
@@ -6,17 +6,39 @@
+  createDirectivesInstances,
+  findDirectiveDefMatches,
+  saveResolvedLocalsInData,
+} from '../instructions/shared';
```
The idea of `packages/core/src/render3/view/elements.ts` file was that it has the logic that is _referenced_ by instructions - by introducing import from the `instructions` package we are creating a circular dep between packages / folders - something I was actually to very much avoid when restructuring runtime.

pkozlowski-opensource commented on packages/core/src/render3/view/elements.ts:
```diff
@@ -6,17 +6,39 @@
+  increaseElementDepthCount,
+  setCurrentTNode,
+  wasLastNodeCreated,
+} from '../state';
```
I would very much want to avoid imports from state in all places apart from instructions - this was the whole point behind pulling the logic out of instructions into the "framework logic" packages / folders that can be referenced from other instructions and parts of the runtime.

crisbeto commented:
[Passing TGP](http://test/OCL:759683665:BASE:759957555:1747480267967:5e362ac8)

crisbeto committed:
refactor(core): consolidate element end instruction logic

There was some identical logic between the `elementEnd` and `elementContainerEnd` instructions. These changes consolidate it.

crisbeto committed:
refactor(core): consolidate logic to determine whether node can be hydrated

Several instructions were repeating the logic that checks if a specific can be hydrated. These changes move it into a common location.

thePunderWoman commented:
This PR was merged into the repository by commit 7a308ccc3bc9764fe4bd73d6c87067b0f8de992d.

The changes were merged into the following branches: main
